Tutorial grammar/capitalization fixes and improved readability (#6233)
thanos-io/thanos: docs/quick-tutorial.md

# Quick Tutorial

Check out the free, in-browser interactive tutorial [Killercoda Thanos course](https://killercoda.com/thanos). We will be progressively updating our Killercoda course with more scenarios.

On top of this, find our quick tutorial below.

## Prometheus

> Thanos is based on Prometheus. With Thanos, Prometheus always remains as an integral foundation for collecting metrics and alerting using local data, but you use Prometheus features to varying degrees depending on the deployment model.

Reviewer: ```suggestion
Thanos is based on Prometheus. With Thanos, Prometheus always remains as an integral foundation for collecting metrics and alerting using local data.
```

> Thanos bases itself on vanilla [Prometheus](https://prometheus.io/) (v2.2.1+). We plan to support *all* Prometheus versions beyond this one.

Reviewer: ```suggestion
Thanos bases itself on vanilla [Prometheus](https://prometheus.io/). We plan to support *all* Prometheus versions beyond v2.2.1.
```

Always make sure to run Prometheus as recommended by the Prometheus team:

* Put Prometheus in the same failure domain. This means in the same network and in the same geographic location as the monitored services.
* Use a persistent disk to persist data across Prometheus restarts.
* Use local compaction for longer retentions.
* Do not change the minimum TSDB block durations.
* Do not scale out Prometheus unless necessary. A single Prometheus instance is already efficient.

We recommend using Thanos when you need to scale out your Prometheus instance.

## Components

Following the [KISS](https://en.wikipedia.org/wiki/KISS_principle) and Unix philosophies, Thanos is comprised of a set of components where each fulfills a specific role.

* Sidecar: connects to Prometheus, reads its data for query and/or uploads it to cloud storage.
* Store Gateway: serves metrics inside of a cloud storage bucket.
* Compactor: compacts, downsamples and applies retention on the data stored in the cloud storage bucket.
* Receiver: receives data from Prometheus's remote write write-ahead log, exposes it, and/or uploads it to cloud storage.

> * Ruler/Rule: evaluates recording and alerting rules against data in Thanos for exposure and/or upload.

Reviewer: "exposition" is a word from the Prometheus domain relating to data format and is the correct word here 
```suggestion
* Ruler/Rule: evaluates recording and alerting rules against data in Thanos for exposition and/or upload.
```

* Querier/Query: implements Prometheus's v1 API to aggregate data from the underlying components.
* Query Frontend: implements Prometheus's v1 API to proxy it to Querier while caching the response and optionally splitting it by queries per day.

> Deployment via Kubernetes sidecar:

Reviewer: Sidecar is another Thanos component and should be left capitalized
```suggestion
Deployment with Thanos Sidecar for Kubernetes:
```

Author: Hmmm, this is an interesting case. I could see keeping it capitalized as an official component of Thanos. But on the other hand, I'm not clear how it differs from the standard concept of a Kubernetes sidecar. It's sort of a judgement call; you guys probably know best so I'm okay to go that route if this way sounds more acceptable to the Thanos community. We'll also want to bring https://github.com/thanos-io/thanos/blob/main/docs/components/sidecar.md into parity soon too then.

But if we go the capitalization route, then we'll generally need to drop `a` and `the` before it, as it will become a proper noun, so some of the other suggestions will need to be revised accordingly.

<!---
Source file to copy and edit: https://docs.google.com/drawings/d/1AiMc1qAjASMbtqL6PNs0r9-ynGoZ9LIAtf0b9PjILxw/edit?usp=sharing
-->

![Sidecar](https://docs.google.com/drawings/d/e/2PACX-1vSJd32gPh8-MC5Ko0-P-v1KQ0Xnxa0qmsVXowtkwVGlczGfVW-Vd415Y6F129zvh3y0vHLBZcJeZEoz/pub?w=960&h=720)

> Deployment via Receive in order to scale out or implement with other remote write compatible sources:

Reviewer: "implement" sounds funny to me
```suggestion
Deployment via Receive in order to scale out or interface with other remote write-compatible sources:
```

### Store API

> The sidecar component implements and exposes a gRPC *[Store API](https://github.com/thanos-io/thanos/blob/main/pkg/store/storepb/rpc.proto#L27)*. The sidecar implementation allows you to query the metric data stored in Prometheus.

Reviewer: ```suggestion
The Sidecar component implements and exposes a gRPC *[Store API](https://github.com/thanos-io/thanos/blob/main/pkg/store/storepb/rpc.proto#L27)*. The sidecar implementation allows you to query the metric data stored in Prometheus.
```

Author: This may be a case where leaving `The` is okay since `Sidecar` is followed by `component`. Bit of a squishy case though, especially since it's followed by a sentence starting with `The sidecar implementation`.

There's also an uncapitalized `the sidecar` in line 65 -- worth changing?

> Let's extend the sidecar in the previous section to connect to a Prometheus server, and expose the Store API:

Reviewer: ```suggestion
Let's extend the Sidecar in the previous section to connect to a Prometheus server, and expose the Store API:
```

* *[Example Kubernetes manifests using Prometheus operator](https://github.com/coreos/prometheus-operator/tree/master/example/thanos)*

### Uploading Old Metrics

> When the sidecar is run with the `--shipper.upload-compacted` flag, it will sync all older existing blocks from Prometheus local storage on startup. 

Reviewer: ```suggestion
When Sidecar is run with the `--shipper.upload-compacted` flag, it will sync all older existing blocks from Prometheus local storage on startup. 
```

> NOTE: This assumes you never run the sidecar with block uploading against this bucket. Otherwise, you must manually remove overlapping blocks from the bucket. Those mitigations will be suggested in the sidecar verification process.

Reviewer: ```suggestion
NOTE: This assumes you never run the Sidecar with block uploading against this bucket. Otherwise, you must manually remove overlapping blocks from the bucket. Those mitigations will be suggested in the sidecar verification process.
```

Author: Same thing with `the`

## Querier/Query

> Now that we have setup the sidecar for one or more Prometheus instances, we want to use Thanos's global [Query Layer](components/query.md) to evaluate PromQL queries against all instances at once.

Reviewer: ```suggestion
Now that we have setup the Sidecar for one or more Prometheus instances, we want to use Thanos's global [Query Layer](components/query.md) to evaluate PromQL queries against all instances at once.
```

Author: Same thing with `the`

> The Querier component is stateless and horizontally scalable, and can be deployed with any number of replicas. Once connected to the sidecars, it automatically detects which Prometheus servers need to be contacted for a given PromQL query.

Reviewer: ```suggestion
The Querier component is stateless and horizontally scalable, and can be deployed with any number of replicas. Once connected to the Sidecars, it automatically detects which Prometheus servers need to be contacted for a given PromQL query.
```

Author: Same thing with `the`, and since this is plural I think it's best to prepend a clarifying `Thanos` to `Sidecars`

Author: Actually the more I look at this the more I feel like I've forgotten how to English... since it's plural, keeping `the` may actually make the most sense. I don't know if there's a formal rule for that, but that way sounds more intuitively natural.

> Thanos Querier also implements Prometheus's official HTTP API and can thus be used with external tools such as Grafana. It also serves a derivative of Prometheus's UI for ad-hoc querying and stores its status.

Reviewer: The intention here is to indicate that users can view that status of the various "Thanos stores" configured in the deployment.
```suggestion
Thanos Querier also implements Prometheus's official HTTP API and can thus be used with external tools such as Grafana. It also serves a derivative of Prometheus's UI for ad-hoc querying and checking the status of the Thanos stores.
```

> Below, we will set up a Thanos Querier to connect to our sidecars, and expose its HTTP UI:

Reviewer: ```suggestion
Below, we will set up a Thanos Querier to connect to our Sidecars, and expose its HTTP UI:
```

    --store        dnssrv+_grpc._tcp.thanos-store.monitoring.svc  # Supports DNS A & SRV records
```

> Go to the configured HTTP address, which should now show a UI similar to that of Prometheus. You can now query across all Prometheus instances within the cluster. You can also check out the Stores page, which shows all of your Stores.

Reviewer: Not sure but think we shouldn't capitalize 'store'
```suggestion
Go to the configured HTTP address, which should now show a UI similar to that of Prometheus. You can now query across all Prometheus instances within the cluster. You can also check out the Stores page, which shows all of your stores.
```

Thanos Compactor simply scans the object storage bucket and performs compaction where required. At the same time, it is responsible for creating downsampled copies of data in order to speed up queries.

```bash
thanos compact \
    --data-dir             /var/thanos/compact \  # Temporary workspace for data processing
    --objstore.config-file bucket_config.yaml \   # Bucket where compacting will be performed
    --http-address         0.0.0.0:19191          # HTTP endpoint for collecting metrics on the compactor
```

Compactor is not in the critical path of querying or data backup. It can either be run as a periodic batch job or be left running to always compact data as soon as possible. It is recommended to provide 100-300GB of local disk space for data processing.

*NOTE: Compactor must be run as a **singleton** and must not run when manually modifying data in the bucket.*

* *[Example Kubernetes manifest](https://github.com/thanos-io/kube-thanos/blob/master/examples/all/manifests/thanos-compact-statefulSet.yaml)*

[Compactor documentation](components/compact.md)

## Ruler/Rule

> In case Prometheus running with a Thanos sidecar does not have enough retention, or if you want to have alerts or recording rules that require a global view, Thanos has just the component for that: the [Ruler](components/rule.md), which does rule and alert evaluation on top of a given Thanos Querier.

Reviewer: ```suggestion
In case Prometheus running with a Thanos Sidecar does not have enough retention, or if you want to have alerts or recording rules that require a global view, Thanos has just the component for that: the [Ruler](components/rule.md), which does rule and alert evaluation on top of a given Thanos Querier.
```
